File: pset6/credit.py
#import cs50
#
#while True:
#    print("Number: ", end="")
#    number = cs50.get_float()
#    if number >= 0:
#        break

# validate input
while True:
    cc = input("Number: ")
    if cc.isdecimal() and int(cc) > 0:
        break

# copy input into variables; lose last digit for int j
i, j = int(cc), int(cc) // 10
x1, x2 = 0, 0
    
# starting from last digit, add every other digit together
# % 10 yields last digit, // 100 "cuts off" last two
while i > 0:
    x1 += i % 10
    i //= 100
    
# starting from penultimate digit (already there), double every other digit, then add those digits together
while j > 0:
    # if result is not a single digit, % 10 yields last digit and // 10 yields the next one
    if (j % 10) * 2 > 8:
        x2 += ((j % 10) * 2) % 10
        x2 += ((j % 10) * 2) // 10
        j //= 100
    else:
        x2 += (j % 10) * 2
        j //= 100
        
checksum = x1 + x2

# if last digit of checksum = 0, and length and first digits are a match, print name
if checksum % 10 == 0:
    if len(cc) == 15 and cc[:2] == "34" or cc[:2] == "37":
        print("AMEX")
    elif len(cc) == 16 and cc[:2] == "51" or cc[:2] == "52" or cc[:2] == "53" or cc[:2] == "54" or cc[:2] == "55":
        print("MASTERCARD")
    elif len(cc) == 13 or len(cc) == 16 and cc[:1] == "4":
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>

long long cc_number, i, j;
int x1, x2, checksum;

int main(void)
{
    do
    {
    printf("Number: ");
    cc_number = get_long_long();
    }
while (cc_number < 0);
    
for (i = cc_number; i > 0; i /= 100)
    x1 += i % 10;

for (j = cc_number / 10; j > 0; j /= 100)
{
    if ((j % 10) * 2 > 8)
    {
        x2 += ((j % 10) * 2) / 10;
        x2 += ((j % 10) * 2) % 10;
    }
    else
    x2 += (j % 10) * 2;
}

checksum = x1 + x2;

if (checksum % 10 == 0)
{
    if (cc_number > 5100000000000000 && cc_number < 5600000000000000)
    printf("MASTERCARD\n");
    else if ((cc_number >=4000000000000 && cc_number < 5000000000000) || (cc_number >=4000000000000000 && cc_number < 5000000000000000))
    printf("VISA\n");
    else if ((cc_number >= 340000000000000 && cc_number < 350000000000000) || (cc_number >= 370000000000000 && cc_number < 380000000000000))
    printf("AMEX\n");
    else
    printf("INVALID\n");
}
else
printf ("INVALID\n");

return 0;
}